server/sonar-server/src/main/java/org/sonar/server/ui/PageRepository.java

```diff
@@ -0,0 +1,126 @@
+import org.sonar.api.Startable;
+import org.sonar.api.server.ServerSide;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.api.web.page.Page;
+import org.sonar.api.web.page.Page.Qualifier;
+import org.sonar.api.web.page.PagesDefinition;
+import org.sonar.core.platform.PluginRepository;
+import org.sonar.core.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static java.util.Comparator.comparing;
+import static java.util.Objects.requireNonNull;
+
+@ServerSide
+public class PageRepository implements Startable {
+  private static final Logger LOGGER = Loggers.get(PageRepository.class);
+  private static final Splitter PAGE_KEY_SPLITTER = Splitter.on("/").omitEmptyStrings();
+
+  private final PluginRepository pluginRepository;
+  private final List<PagesDefinition> definitions;
+  private List<Page> pages;
+
+  public PageRepository(PluginRepository pluginRepository) {
+    this.pluginRepository = pluginRepository;
+    // in case there's no page definition
+    definitions = Collections.emptyList();
+  }
+
+  public PageRepository(PluginRepository pluginRepository, PagesDefinition[] pagesDefinitions) {
+    this.pluginRepository = pluginRepository;
+    definitions = ImmutableList.copyOf(pagesDefinitions);
+  }
+
+  @Override
+  public void start() {
+    ContextImpl context = new ContextImpl();
+    definitions.forEach(definition -> definition.define(context));
+    pages = context.getPages().stream()
+      .filter(isWellFormed())
+      .filter(pluginExists())
+      .sorted(comparing(Page::getKey))
+      .collect(Collectors.toList());
+  }
+
+  @Override
+  public void stop() {
+    // nothing to do
+  }
+
+  public List<Page> getGlobalPages(boolean isAdmin) {
+    return getPages(true, isAdmin, null);
+  }
+
+  public List<Page> getComponentPages(boolean isAdmin, String qualifierKey) {
+    Qualifier qualifier = Qualifier.fromKey(qualifierKey);
+    return qualifier == null ? emptyList() : getPages(false, isAdmin, qualifier);
+  }
+
+  private List<Page> getPages(boolean isGlobal, boolean isAdmin, @Nullable Qualifier qualifier) {
+    return getAllPages().stream()
+      .filter(p -> p.isGlobal() == isGlobal)
+      .filter(p -> p.isAdmin() == isAdmin)
+      .filter(p -> qualifier == null || p.getQualifiers().contains(qualifier))
+      .collect(Collectors.toList());
+  }
+
+  @VisibleForTesting
+  List<Page> getAllPages() {
+    return requireNonNull(pages, "Pages haven't been initialized yet");
+  }
+
+  private Predicate<Page> pluginExists() {
+    return page -> {
+      String plugin = PAGE_KEY_SPLITTER.splitToList(page.getKey()).get(0);
+      boolean pluginExists = pluginRepository.hasPlugin(plugin);
+      if (!pluginExists) {
+        LOGGER.warn("Page '{}' references plugin '{}' that does not exist. Page is ignored.", page.getName(), plugin);
+      }
+
+      return pluginExists;
+    };
+  }
+
+  private static Predicate<Page> isWellFormed() {
```

Comment:
As this check does not rely on environment (installed plugins), it should be moved to Page.Builder.